File: plugin/src/main/java/com/jtprince/bingo/plugin/automarking/itemtrigger/ItemTriggerYaml.java
package com.jtprince.bingo.plugin.automarking.itemtrigger;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.io.InputStream;
import java.util.*;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class ItemTriggerYaml {
    public final Map<String, MatchGroup> itemTriggers;

    @JsonCreator
    public ItemTriggerYaml(@JsonProperty("item_triggers") Map<String, MatchGroup> itemTriggers) {
        this.itemTriggers = itemTriggers;
    }

    public static ItemTriggerYaml fromFile(InputStream yamlFile) throws IOException {
        ObjectMapper yaml = new ObjectMapper(new YAMLFactory());
        // Allow `name` field to be either a String or list of Strings
        yaml.configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true);
        return yaml.readValue(yamlFile, ItemTriggerYaml.class);
    }

    public @Nullable ItemTriggerYaml.MatchGroup get(String goalId) {
        return itemTriggers.get(goalId);
    }

    public static class MatchGroup {
        private final List<Pattern> names;
        public final int unique;
        public final int total;
        public final List<MatchGroup> children;

        @JsonCreator
        public MatchGroup(@JsonProperty("name") List<String> name,
                          @JsonProperty("unique") Integer unique,
                          @JsonProperty("total") Integer total,
                          @JsonProperty("groups") List<MatchGroup> groups) {
            if (name != null) {
                this.names = name.stream().map(Pattern::compile).collect(Collectors.toList());
            } else {
                this.names = Collections.emptyList();
            }
            this.unique = Objects.requireNonNullElse(unique, 1);
            this.total = Objects.requireNonNullElse(total, 1);
            this.children = Objects.requireNonNullElse(groups, Collections.emptyList());
        }

        public boolean nameMatches(String name) {
            return names.stream().anyMatch(p -> p.matcher(name).matches());
        }
    }
}

File: kotlinplugin/src/main/kotlin/com/jtprince/bingo/kplugin/automark/ItemTriggerYaml.kt
package com.jtprince.bingo.kplugin.automark

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.jtprince.bingo.kplugin.BingoPlugin
import com.jtprince.bingo.kplugin.board.SetVariables
import java.io.IOException
import java.io.InputStream
import java.util.logging.Level


class ItemTriggerYaml private constructor(
    @JsonProperty("item_triggers") root: Map<String, MatchGroup>
) {
    private val itemTriggers: Map<String, MatchGroup> = root
    companion object {
        val defaultYaml: ItemTriggerYaml by lazy {
            fromFile(ItemTriggerYaml::class.java.getResourceAsStream("/item_triggers.yml"))
        }

        /**
         * Load a YAML specification from an InputStream.
         */
        fun fromFile(yamlFile: InputStream?): ItemTriggerYaml {
            val yaml = ObjectMapper(YAMLFactory())
            // Allow `name` field to be either a String or list of Strings
            yaml.configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true)
            return try {
                yaml.readValue(yamlFile, ItemTriggerYaml::class.java)
            } catch (e: IOException) {
                BingoPlugin.logger.log(Level.SEVERE, "Could not parse item triggers yaml", e)
                ItemTriggerYaml(emptyMap())  // Return an empty YAML
            }
        }
    }

    val allAutomatedGoals: Set<String>
        get() = itemTriggers.keys

    /**
     * Get the specifications for the Item Trigger that is configured for a given goal ID. If the
     * goal ID is not present in item_triggers.yml, returns null. The return value is an Item Match
     * Group that can be considered the Root Item Match Group for this goal ID.
     */
    operator fun get(goalId: String): MatchGroup? {
        return itemTriggers[goalId]
    }

    /**
     * An Item Match Group is a node in a tree of Item Match Group objects, with the root of this
     * tree belonging to a goal ID. This mechanism is described in detail in README.md.
     */
    class MatchGroup @JsonCreator constructor(
        @JsonProperty("name") names: List<String>?,
        @JsonProperty("unique") unique: String?,
        @JsonProperty("total") total: String?,
        @JsonProperty("groups") children: List<MatchGroup>?
    ) {
        private val names: List<Regex> = names?.map{ s -> Regex(s) } ?: emptyList()
        private val unique: Variable = Variable(unique, 1)
        private val total: Variable = Variable(total, 1)
        internal val children: List<MatchGroup> = children ?: emptyList()

        fun nameMatches(name: String): Boolean {
            return names.any { it.matches(name) }
        }

        fun unique(setVariables: SetVariables): Int {
            return unique.getValue(setVariables)
        }

        fun total(setVariables: SetVariables): Int {
            return total.getValue(setVariables)
        }
    }

    internal class Variable(yamlVarValue: String?, defaultConstant: Int) {
        private val name: String?
        private val constant: Int?

        init {
            if (yamlVarValue?.startsWith("$") == true) {
                name = yamlVarValue.substring(1)
                constant = null
            } else {
                name = null
                constant = yamlVarValue?.toIntOrNull() ?: defaultConstant
            }
        }

        fun getValue(setVariables: SetVariables): Int {
            return constant ?: setVariables.getOrElse(name!!) {
                throw RuntimeException("Unknown variable $name")
            }
        }
    }
}
